lib/client.js

```diff
@@ -490,60 +504,72 @@ MqttClient.prototype.publish = function (topic, message, opts, callback) {
     return this
   }
 
-  packet = {
-    cmd: 'publish',
-    topic: topic,
-    payload: message,
-    qos: opts.qos,
-    retain: opts.retain,
-    messageId: this._nextId(),
-    dup: opts.dup
-  }
-
-  if (options.protocolVersion === 5) {
-    packet.properties = opts.properties
-    if ((!options.properties && packet.properties && packet.properties.topicAlias) || ((opts.properties && options.properties) &&
-      ((opts.properties.topicAlias && options.properties.topicAliasMaximum && opts.properties.topicAlias > options.properties.topicAliasMaximum) ||
-        (!options.properties.topicAliasMaximum && opts.properties.topicAlias)))) {
-      /*
-      if we are don`t setup topic alias or
-      topic alias maximum less than topic alias or
-      server don`t give topic alias maximum,
-      we are removing topic alias from packet
-      */
-      delete packet.properties.topicAlias
+  var that = this
+  var publishProc = function () {
+    var messageId = 0
+    if (opts.qos === 1 || opts.qos === 2) {
+      messageId = that._nextId()
+      if (messageId === null) {
```

Comment:
I see here you're checking that it is null, would packet.messageId be null or would it be undefined?

Reply:
It is null.

Let me explain.
messageId is allocated using messageIdProvider.
See https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/client.js#L188
Note that the original implementation using `_nextId()`. I remained `_nextId()` but it uses `messageIdProvider.allocate()`
See https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/client.js#L1503-L1505

There are two types of messageIdProvider.
One is simply incremental. The initial value is random. It is the original behavior. The messageId could conflict. It never return null (or undefined).
See https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/default-message-id-provider.js#L25-L33
The other is allocating an unique number. 
See https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/unique-message-id-provider.js#L23-L26
If the all available ids are used, then return **null**, not undefined.
Implementation is here;
https://github.com/redboltz/number-allocator/blob/1.0.7/lib/number-allocator.js#L62-L84

See alloc() https://www.npmjs.com/package/number-allocator 
## alloc()
Allocate the first vacant number. The number become occupied status.

Time Complexity O(1)
- return: Number
  - The first vacant number. If all numbers are occupied, return null.

So the messageId in the client.js never be undefined.

I noticed that the comment is misleading.  I will update it.
https://github.com/redboltz/MQTT.js/blob/add_unconflict_message_id_allocation/lib/unique-message-id-provider.js#L21